ARROW-2398: [Rust] Create Builder<T> for building buffers directly in aligned memory

andygrove opened the pull request:
Also adds our first example

andygrove commented:
rustfmt has stopped working on my system, so CI will probably fail until I fix that.

xhochy commented on rust/src/buffer.rs:
```diff
@@ -18,10 +18,74 @@
+    }
+
+    /// Build a Buffer from the existing memory
+    pub fn build(&mut self) -> Buffer<T> {
```
Is there a reason that this is called `build`? In C++ we call this `Finish` and it would be nice to have them consistent. If `build` is a typical name used in Rust, it's probably better to keep it as is.

andygrove commented:
Moved builder into separate file for consistency with C++. Also implemented `Drop` for builder (in case builder was created but finish was not called).

pitrou commented:
@andygrove I think you forgot to add `buffer.rs`.

andygrove commented:
Oops, added. Also ran `rustfmt` again.

andygrove commented:
@maxim-lian Vec is almost perfect, yes. The only problem is that the memory isn't aligned to an 8 or 64 byte boundary as required by the Arrow specification.
